text/0078-merkleized-metadata.md

```diff
@@ -0,0 +1,390 @@
+- `signature_ty`: The signature type used by the chain.
+- `signed_extensions`: FRAME based runtimes can extend the base extrinsic with extra information. This extra information that is put into an extrinsic is called "signed extensions". These extensions offer the runtime developer the possibility to include data directly into the extrinsic, like  `nonce`, `tip`, amongst others. This means that the this data is sent alongside the extrinsic to the runtime. The other possibility these extensions offer is to include extra information only in the signed data that is signed by the sender. This means that this data needs to be known by both sides, the signing side and the verification side. An example for this kind of data is the *genesis hash* that ensures that extrinsics are unique per chain. Another example is the *metadata hash* itself that will also be included in the signed data. The offline wallets need to know which signed extensions are present in the chain and this is communicated to them using this field.
+
+The `SignedExtensionMetadata` provides information about a signed extension:
+
+- `identifier`: The `identifier` of the signed extension. An `identifier` is required to be unique in the Polkadot ecosystem as otherwise extrinsics are maybe built incorrectly.
+- `included_in_extrinsic`: The type that will be included in the extrinsic by this signed extension.
+- `included_in_signed_data`: The type that will be included in the signed data by this signed extension.
+
+### Type Information
+
+As SCALE is not self descriptive like JSON, a decoder always needs to know the format of the type to decode it properly. This is where the type information comes into play. The format of the extrinsic is fixed as described above and `ExtrinsicMetadata`  provides information on which type information is required for which part of the extrinsic. So, offline wallets only need access to the actual type information. It is a requirement that the type information can be chunked into logical pieces to reduce the amount of data that is sent to the offline wallets for decoding the extrinsics. So, the type information is structured in the following way:
+
+```rust
+struct Type {
+    path: Vec<String>,
+    type_def: TypeDef,
+    type_id: Compact<u32>,
+}
+
+enum TypeDef {
+    Composite(Vec<Field>),
+    Enumeration(EnumerationVariant),
+    Sequence(TypeRef),
+    Array(Array),
+    Tuple(Vec<TypeRef>),
+    BitSequence(BitSequence),
+}
+
+struct Field {
+    name: Option<String>,
+    ty: TypeRef,
+    type_name: Option<String>,
+}
+
+struct Array {
+    len: u32,
+    type_param: TypeRef,
+}
+
+struct BitSequence {
+    num_bytes: u8,
+    least_significant_bit_first: bool,
+}
+
+struct EnumerationVariant {
+    name: String,
+    fields: Vec<Field>,
+    index: Compact<u32>,
```

Comment:
This should be a `u8`, right? Or I'm missing something?

Reply:
This was done on purpose. From time to time the idea pops up to change this `index` in an enum to a `Compact`. So, far no one has done this :see_no_evil: 

For most of the enums this will be `u8` any way when using `Compact`.

Reply:
Sounds good👍